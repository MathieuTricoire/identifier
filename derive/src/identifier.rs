extern crate proc_macro;

use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse_quote, DeriveInput};

use crate::internals::{attr::Attrs, Ctxt};

pub fn expand_derive_identifier(input: &DeriveInput) -> Result<TokenStream, Vec<syn::Error>> {
    let cx = Ctxt::new();
    let attrs = Attrs::get(&cx, input);
    cx.check()?;

    let ident = &input.ident;
    let with = attrs.with();
    let params = attrs.params();

    let expanded = quote! {
        const _: () = {
            impl identifier::Identifier for #ident {
                type Id = #ident;
                type ParseError = identifier::ParseError;

                fn generate() -> Self::Id {
                    #ident(#with::generate(#params))
                }

                fn format(&self) -> String {
                    format!("{:032x}", self.0)
                }

                fn parse_str(s: &str) -> Result<Self::Id, Self::ParseError> {
                    let s = s.replace("-", "");
                    if s.len() != 32 {
                        Err(identifier::ParseError::InvalidLength)
                    } else {
                        let value = u128::from_str_radix(&s, 16).or(Err(identifier::ParseError::InvalidChars))?;
                        if (#with::validate(value, #params)) {
                            Ok(#ident(value))
                        } else {
                            Err(identifier::ParseError::Invalid)
                        }
                    }
                }
            }
        };
    };

    Ok(expanded)
}

pub fn expand_derive_display(input: &DeriveInput) -> Result<TokenStream, Vec<syn::Error>> {
    let ident = &input.ident;

    let expanded = quote! {
        impl std::fmt::Display for #ident {
            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(f, "{}", self.format())
            }
        }
    };

    Ok(expanded)
}

pub fn expand_derive_from_str(input: &DeriveInput) -> Result<TokenStream, Vec<syn::Error>> {
    let ident = &input.ident;
    let trait_path: syn::Path = parse_quote!(core::str::FromStr);

    let expanded = quote! {
        impl #trait_path for #ident {
            type Err = identifier::ParseError;

            fn from_str(s: &str) -> Result<Self, Self::Err> {
                #ident::parse_str(s)
            }
        }
    };

    Ok(expanded)
}
