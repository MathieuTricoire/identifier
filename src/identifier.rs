/// Trait for id structures.
///
/// This trait is not intented to be implemented manually you should use derive.
/// Refers to the module documentation to provide a module with functions to:
/// - generate inner id / u128 value
/// - validate
///
pub trait Identifier: std::cmp::Eq {
    type Id;
    type ParseError;

    /// Generate a new id
    fn generate() -> Self::Id;

    /// format the id to a string
    fn format(&self) -> String;

    /// Parse a string to an id
    fn parse_str(s: &str) -> Result<Self::Id, Self::ParseError>;
}
