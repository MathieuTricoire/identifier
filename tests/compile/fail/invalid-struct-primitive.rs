use identifier::Identifier;

mod mod_id {
    pub fn generate() -> u128 {
        1
    }

    pub fn validate(_value: u128) -> bool {
        true
    }
}

#[derive(Identifier, Eq, PartialEq)]
#[identifier(with = "mod_id")]
struct Id(u64);

fn main() {}
