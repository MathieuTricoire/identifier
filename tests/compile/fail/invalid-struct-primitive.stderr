error: Only `u128` primitive type is supported.
  --> $DIR/invalid-struct-primitive.rs:15:11
   |
15 | struct Id(u64);
   |           ^^^
